git-branchless-revset/src/builtins.rs

```diff
@@ -178,8 +180,48 @@ fn fn_heads(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
 
 #[instrument]
 fn fn_branches(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
-    eval0(ctx, name, args)?;
-    Ok(ctx.dag.branch_commits.clone())
+    let pattern = match eval0_or_1_pattern(ctx, name, args)? {
+        Some(pattern) => pattern,
+        None => return Ok(ctx.dag.branch_commits.clone()),
+    };
+
+    let branch_commits = make_pattern_matcher_for_set(
+        ctx,
+        name,
+        args,
+        Box::new(move |repo, commit| {
+            let branch_oid_to_names = match repo.get_references_snapshot() {
```

Comment:
issue: I'm concerned that this means that we'll try to read all references on the repo once for every commit. Instead, can you get the references snapshot just once outside of this closure (and implicitly `move` it in)?

Reply:
Hmm, yes. I think I was concerned that `repo` was being passed to the closure to handle some advanced usage where we may be working with commits from different repos, but that doesn't make sense in hindsight, and extracting it makes the code a lot cleaner too. 👍